core/include/moveit/task_constructor/properties.h

```diff
@@ -101,18 +131,39 @@ class Property
 	/// reset to default value (which can be empty)
 	void reset();
 
-	/// the current value defined or will the fallback be used?
-	inline bool defined() const { return !value_.empty(); }
+	/// is a value defined?
+	inline bool defined() const { return !(value_.empty() && default_.empty()); }
+
+	/// is a non-default value set?
+	inline bool hasCurrentValue() const { return !value_.empty(); }
 
-	/// get current value (or default if not defined)
+	/// get current value (or default if not set)
 	inline const boost::any& value() const { return value_.empty() ? default_ : value_; }
-	inline boost::any& value() { return value_.empty() ? default_ : value_; }
+
+	/// get typed value of property. Throws bad_any_cast on type mismatch, undefined if !defined().
+	template <typename T>
+	inline const T& value() const {
+		const boost::any& v{ value() };
+		if (v.empty())
+			throw Property::undefined();
+		return boost::any_cast<const T&>(value());
```

Comment:
```suggestion
		return boost::any_cast<const T&>(v);
```
See also my comment on `PropertyMap::get()`.